perception/sub8_perception/nodes/buoy_finder.py

```diff
@@ -0,0 +1,364 @@
+#!/usr/bin/env python
+import cv2
+import numpy as np
+import rospy
+from image_geometry import PinholeCameraModel
+import mil_ros_tools
+import tf
+from collections import deque
+from sub8_vision_tools import rviz, MultiObservation
+from sub8_msgs.srv import VisionRequest2DResponse, VisionRequest2D, VisionRequest, VisionRequestResponse
+from std_msgs.msg import Header
+from std_srvs.srv import SetBool, SetBoolResponse
+from geometry_msgs.msg import Pose2D, PoseStamped, Pose, Point
+from mil_ros_tools import Image_Subscriber, Image_Publisher
+
+
+class Buoy(object):
+    '''
+    Represents one colored buoy for use in BuoyFinder. Contains
+    color values to segment an image for its color and an internal
+    buffer of observations to use in position estimation.
+    '''
+    def __init__(self, color, debug_cv=False):
+        self.observations = deque()
+        self.pose_pairs = deque()
+        self.last_t = None
+        self.debug_cv = debug_cv
+        self.status = ''
+        self.est = None
+        self.color = color
+
+        # Only for visualization
+        if color == 'red':
+            self.draw_colors = (1.0, 0.0, 0.0, 1.0)
+            self.cv_colors = (0, 0, 255, 0)
+            self.visual_id = 0
+        elif color == 'green':
+            self.draw_colors = (0.0, 1.0, 0.0, 1.0)
+            self.cv_colors = (0, 255, 0, 0)
+            self.visual_id = 1
+        elif color == 'yellow':
+            self.draw_colors = (1.0, 1.0, 0.0, 1.0)
+            self.cv_colors = (0, 255, 255, 0)
+            self.visual_id = 2
+        else:
+            rospy.logerr('Unknown buoy color {}'.format(color))
+            self.draw_colors = (0.0, 0.0, 0.0, 1.0)
+            self.visual_id = 3
+        if self.debug_cv:
+            cv2.namedWindow(self.color)
+
+    def load_segmentation(self):
+        '''
+        Load threshold values in BGR, HSV, or LAB colorspace for segmenting an image.
+        '''
+        for color_space in ['hsv', 'bgr', 'lab']:
+            self.color_space = color_space
+            low = '/color/buoy/{}/{}_low'.format(self.color, color_space)
+            high = '/color/buoy/{}/{}_high'.format(self.color, color_space)
+            if not rospy.has_param(low):
+                continue
+            self.thresholds = [np.array(rospy.get_param(low)),
+                               np.array(rospy.get_param(high))]
+            rospy.loginfo("BUOY - Thresholds for {} buoy loaded using {} colorspace".format(
+                          self.color, self.color_space))
+            return
```

Comment:
Should this `return` be here? The debug stuff won't work unless there are no color space params.